ios/UARCTModule/UARCTEventEmitter.m

```diff
@@ -36,37 +40,51 @@ - (instancetype)init {
     self = [super init];
     if (self) {
         self.pendingEvents = [NSMutableArray array];
+        self.knownListeners = [NSMutableSet set];
     }
 
     return self;
 }
 
-- (BOOL)sendEventWithName:(NSString *)eventName body:(id)body {
-    if (self.bridge && self.isObserving) {
+- (void)sendEventWithName:(NSString *)eventName body:(id)body {
+    if (self.bridge && self.isObserving && [self.knownListeners containsObject:eventName]) {
         [self.bridge enqueueJSCall:@"RCTDeviceEventEmitter"
                             method:@"emit"
                               args:body ? @[eventName, body] : @[eventName]
                         completion:NULL];
 
-        return YES;
+    } else {
+        @synchronized(self.pendingEvents) {
+            [self.pendingEvents addObject:@{ UARCTEventNameKey: eventName, UARCTEventBodyKey: body}];
+        }
     }
-
-    return NO;
 }
 
 - (void)addListener:(NSString *)eventName {
     self.listenerCount++;
-    if (self.listenerCount > 0) {
-        for (NSDictionary *event in self.pendingEvents) {
-            [self sendEventWithName:event[@"name"] body:event[@"body"]];
-        }
 
-        [self.pendingEvents removeAllObjects];
+    @synchronized(self.knownListeners) {
```

Comment:
I think you need to add the eventName to the listeners after you flush the pending events. As it is, if an event arrives before you are done flushing, that event might be passed to the listener out of order.